sound/soc/sof/control.c

```diff
@@ -11,8 +11,38 @@
 /* Mixer Controls */
 
 #include <linux/pm_runtime.h>
+#include <linux/leds.h>
 #include "sof-priv.h"
 
+static void update_mute_led(struct snd_sof_control *scontrol,
+			    struct snd_kcontrol *kcontrol,
+			    struct snd_ctl_elem_value *ucontrol)
+{
+	unsigned int temp = 0;
+	unsigned int mask;
+	int i;
+
+	mask = 1U << snd_ctl_get_ioffidx(kcontrol, &ucontrol->id);
+
+	temp &= ~mask;
```

Comment:
temp is zero initially, so does this result in something non-zero? 0 & <x> -> 0...

Reply:
so just to show you how this works currently (I'm removing the "temp &= ~mask;" as temp is set to zero... I'm toggling dmic mute from ubuntu UI:

So with 1 toggle I'm always getting 2 update events, I guess that means that snd_sof_switch_put is called twice always, don't know why that is so,  maybe we get calls to both channels(?). 

tmp0 is just temp initialized to 0 and temp is the value set in the actual loop. And you see that mask is always 1, ledv is the current led value stored in scontrol->led struct. So this works, but you see that always 1 of the 2 calls is bailing out because value is already set.

1) mute off:
update mute led
mask: 00000000000000000000000000000001
tem0: 00000000000000000000000000000000
temp: 00000000000000000000000000000001
ledv: 00000000000000000000000000000000
setting led value off

update mute led
mask: 00000000000000000000000000000001
tem0: 00000000000000000000000000000000
temp: 00000000000000000000000000000001
ledv: 00000000000000000000000000000001
led value same as temp

2) mute on
update mute led
mask: 00000000000000000000000000000001
tem0: 00000000000000000000000000000000
temp: 00000000000000000000000000000001
ledv: 00000000000000000000000000000001
led value same as temp

update mute led
mask: 00000000000000000000000000000001
tem0: 00000000000000000000000000000000
ledv: 00000000000000000000000000000001
setting led value on